compiler/src/org.graalvm.compiler.hotspot.aarch64.test/src/org/graalvm/compiler/hotspot/aarch64/test/UseBarriersForVolatileTest.java

```diff
@@ -0,0 +1,296 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.graalvm.compiler.hotspot.aarch64.test;
+
+import jdk.vm.ci.aarch64.AArch64;
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+import org.graalvm.compiler.api.test.Graal;
+import org.graalvm.compiler.core.test.GraalCompilerTest;
+import org.graalvm.compiler.graph.Node;
+import org.graalvm.compiler.hotspot.HotSpotGraalRuntimeProvider;
+import org.graalvm.compiler.nodes.StartNode;
+import org.graalvm.compiler.nodes.StructuredGraph;
+import org.graalvm.compiler.nodes.extended.MembarNode;
+import org.graalvm.compiler.nodes.memory.MemoryCheckpoint;
+import org.graalvm.compiler.nodes.memory.ReadNode;
+import org.graalvm.compiler.nodes.memory.WriteNode;
+import org.graalvm.compiler.runtime.RuntimeProvider;
+import org.graalvm.compiler.serviceprovider.GraalServices;
+import org.graalvm.compiler.test.SubprocessUtil;
+import org.graalvm.word.LocationIdentity;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.graalvm.compiler.test.SubprocessUtil.getProcessCommandLine;
+import static org.graalvm.compiler.test.SubprocessUtil.withoutDebuggerArguments;
+
+/**
+ * A test class to exercise AArch64 volatile reads and writes in a subordinate JVM where OpenJDK
+ * config setting UseBarriersForVolatile is set to true.
+ *
+ * The single test in this class creates a new JVM to run only the tests in a specific secondary
+ * test class. It uses the same JVM configuration, test runner and test configuration that it was
+ * started with, modulo a small set of options added to the command line. The important changes are
+ * to reset the OpenJDK option UseBarriersForVolatile from the default value false to true amd to
+ * replace the original list of test classes with the secondary test class.
+ *
+ * The secondary test class is actually a static inner class of this class. This presents the
+ * problem that test annotations on the inner class make it eligible for execution in the original
+ * JVM as well as in the subordinate JVM. This problem is finessed by defining a system property
+ * unique to each run and makeing the test succeed without executing when the property is not set.
+ * Th e relevant system property is not setin the outer JVM but is set by the command line used to
```

Comment:
Something strange happened with spaces on this line (`s/Th e/The/`, `s/setin/set in/`).

Regarding the approach with the system property i think it makes sense :+1:.

Reply:
This comment is actually out of date. Originally I just used a system property based on the test name. I tried changing it to using the current time to define the property name so that there was little chance of ti being set on the test command line by mistake. Of course, that didn't work because I could not guarantee that the property was also set to the same value in the subordinate JVM. So, I have reverted to defining the property name based on the test class name. I'll update the comment to correct this.